src/compiler/checker.ts

```diff
@@ -6710,7 +6712,8 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 return visitAndTransformType(type, type => conditionalTypeToTypeNode(type as ConditionalType));
             }
             if (type.flags & TypeFlags.Substitution) {
-                return typeToTypeNodeHelper((type as SubstitutionType).baseType, context);
+                const typeNode = typeToTypeNodeHelper((type as SubstitutionType).baseType, context);
+                return isNoInferType(type) ? factory.createTypeReferenceNode("NoInfer", [typeNode]) : typeNode;

```

Comment:
What happens in declaration emit when something like this gets serialized?

```ts
// foo.ts
export const f: <T>(x: T, y: NoInfer<T>) => bool;

// bar.ts

import { f } from "./foo.js";

type NoInfer<T> = number;

export const g = f;
```

Reply:
With the latest commit this will now emit `globalThis.NoInfer<T>`, similar to what we do for other global types in conflict situations.

Reply:
Do we have a test for that? Same with Uppercase etc.

Reply:
We don't, at least not that I can tell.

Reply:
Test added.